plugin/trino-base-jdbc/src/test/java/io/trino/plugin/jdbc/TestJdbcConnectionCreation.java

```diff
@@ -0,0 +1,163 @@
+import com.google.inject.Provides;
+import com.google.inject.Singleton;
+import io.trino.plugin.jdbc.credential.CredentialProvider;
+import io.trino.plugin.jdbc.mapping.IdentifierMapping;
+import io.trino.spi.connector.ConnectorSession;
+import io.trino.testing.AbstractTestQueryFramework;
+import io.trino.testing.QueryRunner;
+import org.h2.Driver;
+import org.testng.annotations.BeforeMethod;
+import org.testng.annotations.Test;
+
+import java.sql.Connection;
+import java.sql.SQLException;
+import java.util.Map;
+import java.util.Optional;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.atomic.AtomicInteger;
+
+import static io.trino.plugin.jdbc.H2QueryRunner.createH2QueryRunner;
+import static io.trino.plugin.jdbc.TestingH2JdbcModule.createProperties;
+import static io.trino.tpch.TpchTable.NATION;
+import static io.trino.tpch.TpchTable.REGION;
+import static org.assertj.core.api.Assertions.assertThat;
+import static org.assertj.core.api.Assertions.fail;
+
+@Test(singleThreaded = true)
+public class TestJdbcConnectionCreation
+        extends AbstractTestQueryFramework
+{
+    private final Map<Connection, Exception> connectionCreations = new ConcurrentHashMap<>();
+    private final AtomicInteger openConnectionCount = new AtomicInteger();
+    private final AtomicInteger closedConnectionCount = new AtomicInteger();
+
+    @BeforeMethod
+    public void verifyAllConnectionsAreClosed()
+    {
+        Optional<Exception> leakedConnection = connectionCreations.values()
+                .stream()
+                .findFirst();
+        if (leakedConnection.isPresent()) {
+            fail("Connection leaked", leakedConnection.get());
+        }
+        assertThat(openConnectionCount.get()).isEqualTo(closedConnectionCount.get());
+    }
+
+    @Override
+    protected QueryRunner createQueryRunner()
+            throws Exception
+    {
+        return createH2QueryRunner(ImmutableList.of(NATION, REGION), createProperties(), new ReusingConnectionH2Module());
+    }
+
+    @Test
+    public void testOpenConnectionsForShowQueries()
+    {
+        assertOpenConnections("SHOW SCHEMAS", 1);
+        assertOpenConnections("SHOW TABLES", 2);
+    }
+
+    @Test
+    public void testOpenConnectionsForTableQueries()
+    {
+        assertOpenConnections("SELECT * FROM nation", 3);
+        assertOpenConnections("CREATE TABLE copy_of_nation AS SELECT * FROM nation", 10);
+        assertOpenConnections("INSERT INTO copy_of_nation SELECT * FROM nation", 11);
+        assertOpenConnections("DROP TABLE copy_of_nation", 3);
+    }
+
+    @Test
+    public void testOpenConnectionsForInformationSchemas()
+    {
+        assertOpenConnections("SELECT * FROM information_schema.schemata", 1);
+        assertOpenConnections("SELECT * FROM information_schema.tables", 1);
+        assertOpenConnections("SELECT * FROM information_schema.columns", 5);
+    }
+
+    @Test
+    public void testOpenConnectionsForPushdownQueries()
```

Comment:
Add separate test coverage for planning and execution

eg
```
assertOpenConnections(String sql, int expectedPlanningConections, int expectedExecutionConnections)
```

this would do
```
assertOpenConnections("EXPLAIN" + sql, expectedPlanningConections);
assertOpenConnections(sql, expectedPlanningConections + expectedExecutionConnections);
```

-- this would help understand why we're not reaching 1 connection per query, even with reuse